[ WIP ] SepTop Protocol

hannahbaumann opened the pull request:
Checklist
* [ ] Added a ``news`` entry

## Developers certificate of origin
- [ ] I certify that this contribution is covered by the MIT License [here](https://github.com/OpenFreeEnergy/openfe/blob/main/LICENSE) and the **Developer Certificate of Origin** at <https://developercertificate.org/>.

IAlibay commented on openfe/protocols/openmm_md/plain_md_methods.py:
```diff
@@ -431,26 +431,29 @@ def _run_MD(simulation: openmm.app.Simulation,
-            density=True,
-            speed=True,
-        ))
+        if output_settings.production_trajectory_filename:
```
@hannahbaumann - I was thinking, do you want to just make this into a separate PR? I feel like we could merge this right now with little to no review.

hannahbaumann replied:
Will do!

IAlibay commented on openfe/protocols/openmm_septop/base.py:
```diff
@@ -0,0 +1,1515 @@
+            verbose=self.verbose,
+            shared_basepath=self.shared_basepath,
+        )
+        state = simulation.context.getState(getPositions=True)
```
I've found that passing enforcePeriodicBox=True here will make things less likely to explode on you later on

hannahbaumann replied:
Thanks, I now added that here as well!

hannahbaumann commented on docs/guide/protocols/septop.rst:
```diff
@@ -0,0 +1,129 @@
+
+Molecular interactions are modified during an alchemical path using a discrete set of lambda windows.
+For the transformation of ligand A to ligand B in the binding site, ligand A is initially fully interacting while ligand B is decoupled.
+First, the non-interacting dummy ligand B is inserted into the binding site and restrained using
```
Changed this to a list!

hannahbaumann commented on docs/guide/protocols/septop.rst:
```diff
@@ -0,0 +1,129 @@
+The :class:`SepTopProtocol <.SepTopProtocol>` calculates the difference in binding free energy between two ligands.
+This protocol essentially performs two absolute binding free energy calculations simultaneously in opposite directions,
+by (alchemically) inserting one ligand into the binding site, while removing the other ligand at the same time.
+In contrast to the :class:`.RelativeHybridTopologyProtocol`, the two ligand topologies are
```
Done!

jthorton commented on docs/guide/protocols/septop.rst:
```diff
@@ -0,0 +1,130 @@
+~~~~~~~~~~~~~~~~~~~
+
+The :class:`.ProtocolDAG` of the :class:`SepTopProtocol <.SepTopProtocol>` contains :class:`.ProtocolUnit`\ s from both the complex and solvent transformations.
+This means that both legs of the thermodynamic cycle are constructed and run concurrently in the same :class:`.ProtocolDAG`. This is different from the :class:`.RelativeHybridTopologyProtocol` where the :class:`.ProtocolDAG` only runs a single leg of a thermodynamic cycle.
```
Should this be sequentially rather than concurrently, or do we currently run them at the same time without the need for exorcist?

hannahbaumann replied:
Great point, will change this!

jthorton commented:
Docs look awesome, great job @hannahbaumann!

IAlibay commented on openfe/protocols/openmm_septop/femto_utils.py:
```diff
@@ -0,0 +1,63 @@
+    positions: openmm.unit.Quantity,
+    box_vectors: openmm.unit.Quantity | None,
+    context_params: dict[str, float] | None = None,
+    platform: OpenMMPlatform = OpenMMPlatform.REFERENCE,
```
This doesn't get used anywhere right now.

IAlibay replied:
I would remove the import here and just have folks pass the platforms themselves like in: https://github.com/choderalab/openmmtools/blob/59c14bfc147a4128eebd61f75927875c2b9976ad/openmmtools/tests/test_alchemy.py#L81

IAlibay commented on openfe/protocols/openmm_septop/femto_utils.py:
```diff
@@ -0,0 +1,63 @@
+def compute_energy(
+    system: openmm.System,
+    positions: openmm.unit.Quantity,
+    box_vectors: openmm.unit.Quantity | None,
```
This isn't used but might be worth keeping around.

IAlibay reviewed (changes requested):
Ok here were some notes I had lying around for the femto vendors.

Overall - `femto_constants` doesn't seem like it's worth keeping around, I would just remove it.

The `is_close` method, it would be best if we used `assert_allclose` like we use elsewhere.

The `compute_energy` method does have advantages over the openmmtools one, but also has some disadvantages (like not being able to pass a platform through directly). My suggestion is we create our own version (put it in the protocol tests conftest maybe?), which does more of what we need.

hannahbaumann commented on openfe/protocols/openmm_septop/femto_constants.py:
```diff
@@ -0,0 +1,49 @@
+
+    COM_RESTRAINT = "com-restraint"
+    POSITION_RESTRAINT = "position-restraint"
+    ALIGNMENT_RESTRAINT = "alignment-restraint"
```
Ah yes, looks like I forgot to remove this after switching to the other alchemical factory, removed the entire file.

hannahbaumann commented on openfe/tests/protocols/test_openmm_septop_slow.py:
```diff
@@ -0,0 +1,310 @@
+    SepTopSolventSetupUnit,
+    SepTopProtocol,
+)
+from openfe.protocols.openmm_septop.femto_utils import compute_energy, is_close
```
Removed this.

hannahbaumann commented on openfe/protocols/openmm_septop/femto_utils.py:
```diff
@@ -0,0 +1,63 @@
+
+from .femto_constants import OpenMMPlatform
+
+def is_close(
```
switched this to `assert_all_close` and `assert not np.assert_close()` when it's not supposed to be equal. Not sure if there is a better way to do the assert not?!

IAlibay replied:
That's the way to do it as far as I'm aware.

IAlibay commented on openfe/protocols/openmm_septop/base.py:
```diff
@@ -0,0 +1,1436 @@
+        if solvent_comp is not None:
+            constants['pressure'] = ensure_quantity(pressure, 'openmm')
+
+        # Get the composable states
+        composable_states = [
+            alchemical_state
+        ]
```
Nitpick:
I would just move this into the call like here: https://github.com/OpenFreeEnergy/openfe/blob/main/openfe/protocols/openmm_afe/base.py#L678

That way it's easier to diff the two and merge things later.

IAlibay commented on openfe/protocols/openmm_septop/base.py:
```diff
@@ -0,0 +1,1436 @@
+        sampler_context_cache : openmmtools.cache.ContextCache
+          The sampler state context cache.
+        """
+        platform = omm_compute.get_openmm_platform(
```
I would copy over the restrict_cpu code from here: https://github.com/OpenFreeEnergy/openfe/blob/59811f647178f583133fb62ef97da8ae682026e4/openfe/protocols/openmm_afe/base.py#L792

IAlibay commented on openfe/protocols/openmm_septop/base.py:
```diff
@@ -0,0 +1,1436 @@
+
+        integrator = openmmtools.mcmc.LangevinDynamicsMove(
+            timestep=to_openmm(integrator_settings.timestep),
+            collision_rate=to_openmm(
+                integrator_settings.langevin_collision_rate),
```
```suggestion
            collision_rate=to_openmm(integrator_settings.langevin_collision_rate),
```

IAlibay commented on openfe/protocols/openmm_septop/base.py:
```diff
@@ -0,0 +1,1436 @@
+            sampler: multistate.MultiStateSampler,
+            reporter: multistate.MultiStateReporter,
+            settings: dict[str, SettingsBaseModel],
+            dry: bool,
+            verbose: bool,
+            shared_basepath: pathlib.Path,
```
Any reason for passing these directly rather than using them from self?

IAlibay commented on openfe/protocols/openmm_septop/base.py:
```diff
@@ -0,0 +1,1436 @@
+        Returns
+        -------
+        dict : dict[str, str]
+            Dictionary with paths to ...
```
Needs completing

IAlibay commented on openfe/protocols/openmm_septop/base.py:
```diff
@@ -0,0 +1,1436 @@
+          Path to the serialized OpenMM system
+        serialized_topology: pathlib.Path
+          Path to the serialized topology of the system
+        dry: bool
```
missing scratch_basepath and shared_basepath

IAlibay commented on openfe/protocols/openmm_septop/base.py:
```diff
@@ -0,0 +1,1436 @@
+        settings : dict[str, SettingsBaseModel]
+          The dictionary of settings for the protocol.
+        dry : bool
+          Whether or not to dry run the simulation
```
If you do need to pass in verbose / shared_basepath, then these should be reflected in the docstring.

hannahbaumann commented on docs/guide/protocols/septop.rst:
```diff
@@ -0,0 +1,130 @@
+Overview
+--------
+
+The :class:`SepTopProtocol <.SepTopProtocol>` calculates the difference in binding free energy between two ligands.
```
Added the two references here.

IAlibay commented on openfe/protocols/openmm_septop/base.py:
```diff
@@ -0,0 +1,1436 @@
+        alchem_comps, solv_comp, prot_comp, smc_comps = self._get_components()
+
+        system = deserialize(serialized_system)
+        pdb = simtk.openmm.app.pdbfile.PDBFile(str(serialized_topology))
```
```suggestion
        pdb = openmm.app.pdbfile.PDBFile(str(serialized_topology))
```
We shouldn't be using simtk

IAlibay commented on openfe/protocols/openmm_septop/base.py:
```diff
@@ -0,0 +1,1436 @@
+        positions = pdb.getPositions(asNumpy=True)
+
+        # Check that the restraints are correctly applied by running a short equilibration
+        equ_positions, box_AB = self._pre_equilibrate(
```
```suggestion
        equil_positions, box_AB = self._pre_equilibrate(
```
or "equilibrated" (equ is hard to follow name-wise)

hannahbaumann commented on openfe/protocols/openmm_septop/base.py:
```diff
@@ -0,0 +1,1436 @@
+
+        return system_modeller, comp_resids
+
+    def _get_omm_objects(
```
The reason why I split this function into two is that creating the system for the single end states (A and B) is different then creating the system that has combined both ligand A and ligand B. Maybe there's a better way that I'm not seeing.

hannahbaumann commented on openfe/protocols/openmm_septop/base.py:
```diff
@@ -0,0 +1,1436 @@
+        smc_comps_AB: dict[SmallMoleculeComponent, OFFMolecule],
+        smc_off_B: dict[SmallMoleculeComponent, OFFMolecule],
+        settings: dict[str, SettingsBaseModel],
+        shared_basepath: pathlib.Path,
```
Unused, removed this.

hannahbaumann commented on openfe/protocols/openmm_septop/equil_septop_method.py:
```diff
@@ -0,0 +1,2135 @@
+        mdtraj_complex_A = _get_mdtraj_from_openmm(omm_topology_A, positions_A)
+        mdtraj_complex_B = _get_mdtraj_from_openmm(omm_topology_B, positions_B)
+        mdtraj_complex_B.superpose(mdtraj_complex_A,
+                                   atom_indices=mdtraj_complex_A.topology.select(
```
This is now in its own method and raises an Error if less than three atom indices are found. 3 is kind of arbritrary, I could also set it to a different number!

hannahbaumann commented on openfe/tests/data/openmm_septop/bace.pdb:
I ended up switching to eg5 system to have less files and to also test a system with cofactor.

hannahbaumann commented on openfe/tests/data/openmm_septop/bace1.sdf:
I ended up switching to eg5 system to have less files and to also test a system with cofactor.

hannahbaumann commented on openfe/protocols/openmm_septop/equil_septop_method.py:
```diff
@@ -0,0 +1,2135 @@
+            ),
+            complex_restraint_settings=BoreschRestraintSettings(
+                K_r=8368.0 * unit.kilojoule_per_mole / unit.nanometer ** 2,
+                K_thetaA=1000 * unit.kilojoule_per_mole / unit.radian ** 2
```
In my previous work I had restrained this angle even stronger than 1000. In initial ABFE benchmarks (where this was set to 83.68) the ligand moved a lot, but there I hadn't yet plotted the distribution. Looking at the recent plots where it is set to 1000, we may want to further increase the restraint strength as the distribution in the restrained, non-interacting state is much wider than in the unrestrained, fully interacting state.

jthorton commented on openfe/protocols/openmm_rfe/equil_rfe_methods.py:
```diff
@@ -703,7 +703,7 @@ def run(self, *, dry=False, verbose=True,
         protocol_settings: RelativeHybridTopologyProtocolSettings = self._inputs['protocol'].settings
         stateA = self._inputs['stateA']
-        stateB = self._inputs['stateB']
+        stateB = self._inputs['stateB']  # TODO: open an issue about this not being used.
```
Has this been done? I would add the issue link here if so.

jthorton commented on openfe/protocols/openmm_septop/base.py:
```diff
@@ -0,0 +1,1300 @@
+from typing import Any
+import openmmtools
+import mdtraj as mdt
+from rdkit import Chem
+from gufe import (
+    ChemicalSystem, SmallMoleculeComponent,
+    ProteinComponent, SolventComponent
+)
+from openfe.protocols.openmm_utils.omm_settings import (
+    SettingsBaseModel,
+)
+from openfe.protocols.openmm_utils.omm_settings import (
+    BasePartialChargeSettings,
+)
+from openfe.utils import log_system_probe
+
+from openfe.protocols.openmm_afe.equil_afe_settings import (
+    BaseSolvationSettings,
+    MultiStateSimulationSettings, OpenMMEngineSettings,
+    IntegratorSettings, LambdaSettings, MultiStateOutputSettings,
+    ThermoSettings, OpenFFPartialChargeSettings, OpenMMSystemGeneratorFFSettings,
+)
+from openfe.protocols.openmm_septop.equil_septop_settings import SepTopSettings
+from openfe.protocols.openmm_utils import omm_compute
+from openfe.protocols.openmm_md.plain_md_methods import PlainMDProtocolUnit
+from ..openmm_utils import (
+    settings_validation, system_creation,
+    multistate_analysis, charge_generation
+)
+from openfe.utils import without_oechem_backend
+from openfe.protocols.restraint_utils.openmm import omm_restraints
+
+from .utils import serialize, deserialize, SepTopParameterState
+from openmmtools.alchemy import (
+    AbsoluteAlchemicalFactory,
+    AlchemicalRegion,
+    AlchemicalState,
+)
+import MDAnalysis as mda
+from openfe.protocols.restraint_utils.geometry.boresch import find_boresch_restraint
```
Woow thats a lot of imports! Can we run isort to make this more readable and group the imports.

hannahbaumann replied:
Sorted and removed unused imports.

jthorton commented on openfe/protocols/openmm_septop/base.py:
```diff
@@ -0,0 +1,1300 @@
+      * `integrator_settings`
+      * `equil_simulation_settings`
+      * `equil_output_settings`
+    endstate: str
+      The endstate that is pre_equilibrates, either 'A' or 'B'.
```
```suggestion
    endstate: str
      The endstate that is pre-equilibrated,can be 'A', 'B' or 'AB'.
```

jthorton commented on openfe/protocols/openmm_septop/base.py:
```diff
@@ -0,0 +1,1300 @@
+        smc_comps_B = smc_off_B | smc_off_both
+        smc_comps_AB = smc_off_A | smc_off_B | smc_off_both
+
+        return smc_comps_A, smc_comps_B, smc_comps_AB, smc_off_B
```
why do we return `smc_off_B` as well?

hannahbaumann replied:
We need it to create the system AB, as we only want the ligand itself for that, not potential cofactors. But since it's only used ones, maybe I can change it to generate it on the fly when needed from the alchemical components dictionary!

jthorton commented on openfe/protocols/openmm_septop/base.py:
```diff
@@ -0,0 +1,1300 @@
+    Base class for running ligand SepTop RBFE free energy transformations.
+    """
+
+    def _prepare(
```
This looks to be the same function in the `BaseSepTopSetupUnit` could we use a mixin?

IAlibay replied:
Might not be worth overly optimizing on this one - in "the refactor" they will both inherit this from some base unit that implements prepare (since it's really needed everywhere).

jthorton commented on openfe/protocols/openmm_septop/base.py:
```diff
@@ -0,0 +1,1300 @@
+                   self.shared_basepath / settings[
+                       'output_settings'].checkpoint_storage_filename]
+            for fn in fns:
+                os.remove(fn)
```
Not critical but it looks like os is imported just for this, we could just do this with pathlib and remove the import.

jthorton commented on openfe/protocols/openmm_septop/equil_septop_method.py:
```diff
@@ -0,0 +1,2125 @@
+
+        return alchem_comps, solv_comp, None, small_mols
+
+    def _handle_settings(self) -> dict[str, SettingsBaseModel]:
```
Again why, is this easier than just using the settings object?

jthorton commented on openfe/protocols/openmm_septop/equil_septop_method.py:
```diff
@@ -0,0 +1,2125 @@
+
+        return production_lengths
+
+    def restraint_geometries(self) -> tuple[list[dict[str, Any]], list[dict[str, Any]]]:
```
Could we return the `BoreschRestraintGeometry` objects they are probably easier to work with and should contain all the information we need?

IAlibay replied:
I _think_ we discussed this in the past, the issue at the moment is that RestraintGeometry isn't a GufeTokenizable. It probably should be, but it might be better for us to deal with this as a separate issue.

jthorton replied:
Is it worth just adding to the restraint API PR?

IAlibay replied:
Separate issue please, turning it into a gufetokenizable is non negligible amounts of work and I'd prefer we deal with it after merge.

jthorton commented on openfe/protocols/openmm_septop/equil_septop_method.py:
```diff
@@ -0,0 +1,2125 @@
+            raise ValueError(errmsg)
+
+        # Crash out if any of the alchemical components are not
+        # SmallMoleculeComponent
+        alchem_components_states = [alchemical_components['stateA'], alchemical_components['stateB']]
+        for state in alchem_components_states:
+            for comp in state:
+                if not isinstance(comp, SmallMoleculeComponent):
+                    errmsg = ("Non SmallMoleculeComponent alchemical species "
+                              "are not currently supported")
+                    raise ValueError(errmsg)
```
I think we can skip making the alchem_components_states and use something like
```python
for state in ["stateA", "stateB"]:
    for comp in alchemical_components[state]:
        if not isinstance(comp, SmallMoleculeComponent):
            raise ValueError("Only SmallMoleculeComponent alchemical species are supported.")
```

hannahbaumann replied:
Thanks! Changed this!

jthorton commented on openfe/protocols/openmm_septop/equil_septop_method.py:
```diff
@@ -0,0 +1,2125 @@
+        lambda_components = [lambda_vdw_A, lambda_vdw_B,
+                             lambda_elec_A, lambda_elec_B,
+                             lambda_restraints_A, lambda_restraints_B]
+        it = iter(lambda_components)
+        the_len = len(next(it))
+        if not all(len(l) == the_len for l in it):
```
We can just use a set to get the same thing I think
```python
lengths = {len(l) for l in lambda_components}
if len(lengths) != 1:
    raise ...
```

jthorton commented on openfe/protocols/openmm_septop/equil_septop_method.py:
```diff
@@ -0,0 +1,2125 @@
+        )
+
+        # Make sure that we have the full system for restraint trajectory analysis
+        if self.settings.complex_equil_output_settings.output_indices != "all":
```
If this is hard set to be all, we probably should not expose it as an option or ignore what the user says, I think that would lead to a better user experience!

IAlibay replied:
This does need rethinking, but given it would require changing a ton of settings, I would suggest punting this to a separate issue.

IAlibay replied:
We ideally need something that will work for both ABFEs and SepTop. Right now I don't have good insight into what that is. I would say this is something we can do as a separate issue that we can do before we release v1.6.

jthorton commented on openfe/protocols/openmm_septop/equil_septop_method.py:
```diff
@@ -0,0 +1,2135 @@
+                stateA=stateA,
+                stateB=stateB,
+                alchemical_components=alchem_comps,
+                generation=0, repeat_id=int(uuid.uuid4()),
```
A mixture of both might make sense, you want some group identifier which you can use to gather all units from a repeat and then some other unique id just incase?

IAlibay replied:
I'm going to reping @atravitz here

jthorton commented on openfe/protocols/openmm_septop/equil_septop_method.py:
```diff
@@ -0,0 +1,2125 @@
+        solvent_run = [
+            SepTopSolventRunUnit(
+                protocol=self,
+                stateA=stateA,
+                stateB=stateB,
+                alchemical_components=alchem_comps,
+                setup=solvent_setup[i],
+                generation=0, repeat_id=int(uuid.uuid4()),
+                name=(f"SepTop RBFE Run, transformation {alchname_A} to "
+                      f"{alchname_B}, solvent leg: repeat {i} generation 0"),
+            )
+            for i in range(self.settings.protocol_repeats)
+        ]
+
+        complex_setup = [
+            SepTopComplexSetupUnit(
+                protocol=self,
+                stateA=stateA,
+                stateB=stateB,
+                alchemical_components=alchem_comps,
+                generation=0, repeat_id=int(uuid.uuid4()),
+                name=(f"SepTop RBFE Setup, transformation {alchname_A} to "
+                      f"{alchname_B}, complex leg: repeat {i} generation 0"),
+            )
+            for i in range(self.settings.protocol_repeats)
+        ]
+
+        complex_run = [
+            SepTopComplexRunUnit(
+                protocol=self,
+                stateA=stateA,
+                stateB=stateB,
+                alchemical_components=alchem_comps,
+                setup=complex_setup[i],
+                generation=0, repeat_id=int(uuid.uuid4()),
+                name=(f"SepTop RBFE Run, transformation {alchname_A} to "
+                      f"{alchname_B}, complex leg: repeat {i} generation 0"),
+            )
+            for i in range(self.settings.protocol_repeats)
+        ]
```
I had a go at trying to reduce the repeated code in this section, it might not be as clear so feel free to leave it
```python
def create_units(unit_cls, leg, setup=None):
    return [
        unit_cls(
            protocol=self,
            stateA=stateA,
            stateB=stateB,
            alchemical_components=alchem_comps,
            setup=setup[i] if setup else None,
            generation=0,
            repeat_id=int(uuid.uuid4()),
            name=(
                f"SepTop RBFE {'Run' if setup else 'Setup'}, transformation {alchname_A} to "
                f"{alchname_B}, {leg} leg: repeat {i} generation 0"
            ),
        )
        for i in range(self.settings.protocol_repeats)
    ]

alchname_A = alchem_comps['stateA'][0].name
alchname_B = alchem_comps['stateB'][0].name

solvent_setup = create_units(SepTopSolventSetupUnit, "solvent")
solvent_run = create_units(SepTopSolventRunUnit, "solvent", setup=solvent_setup)
complex_setup = create_units(SepTopComplexSetupUnit, "complex")
complex_run = create_units(SepTopComplexRunUnit, "complex", setup=complex_setup)
```

hannahbaumann replied:
I was running into errors with this as the SetupUnits don't expect the setup argument. For now I decided to not change the Units themselves and created two separate functions, one for the setup and one for run units. Please let me know if you have a better idea!

jthorton commented on openfe/tests/protocols/openmm_septop/test_septop_protocol.py:
```diff
@@ -0,0 +1,1261 @@
+    with pytest.raises(ValueError, match=errmsg):
+        lambda_settings.lambda_elec_A = val['elec']
+        lambda_settings.lambda_vdw_A = val['vdw']
+        lambda_settings.lambda_restraints_A = val['restraints']
```
I think the cases on lines 75 and 76 need to be worked into the parameterize list, when the error is raised on 74 it wont test the other cases.

jthorton commented on openfe/tests/protocols/openmm_septop/test_septop_protocol.py:
```diff
@@ -0,0 +1,1261 @@
+
+    with pytest.raises(ValueError, match=errmsg):
+        lambda_settings.lambda_elec_A = val['elec']
+        lambda_settings.lambda_vdw_A = val['vdw']
+        lambda_settings.lambda_restraints_A = val['restraints']
```
Same here these lines are skipped.

jthorton commented on openfe/tests/protocols/openmm_septop/test_septop_protocol.py:
```diff
@@ -0,0 +1,1261 @@
+
+def test_create_default_protocol(default_settings):
+    # this is roughly how it should be created
+    protocol = SepTopProtocol(
```
Should we be checking if the settings are preserved as well?

jthorton commented on openfe/tests/protocols/openmm_septop/test_septop_protocol.py:
```diff
@@ -0,0 +1,1261 @@
+def test_charge_error_create(
+    charged_benzene_modifications, T4_protein_component, default_settings
+):
+    # if we create two dags each with 3 repeats, they should give 6 repeat_ids
```
I think this is copied from the other test above and should be removed.

jthorton commented on openfe/tests/protocols/openmm_septop/test_septop_protocol.py:
```diff
@@ -0,0 +1,1261 @@
+        )
+
+
+def test_validate_complex_endstates_protcomp_stateA(
```
Not sure what the preferred pattern is, but for these pairs of tests that test validation in state A and B it would reduce the amount of repeated code if we could use pytest parameterized on them.

jthorton commented on openfe/tests/protocols/openmm_septop/test_septop_slow.py:
```diff
@@ -0,0 +1,333 @@
+        'solvent': SolventComponent(),
+    })
+
+    # Create DAG from protocol, get the vacuum and solvent units
+    # and eventually dry run the first vacuum unit
```
Why do we have a vacuum unit, shouldn't it be complex and solvent only?

hannahbaumann replied:
Changed the comment.

IAlibay commented on openfe/protocols/openmm_septop/base.py:
```diff
@@ -0,0 +1,1300 @@
+                    'simulation_settings'].minimization_steps
+            )
+            # equilibrate
+            if self.verbose:
+                self.logger.info("equilibrating systems")
+
+            sampler.equilibrate(int(equil_steps / mc_steps))  # type: ignore
+
+            # production
+            if self.verbose:
+                self.logger.info("running production phase")
+            sampler.extend(int(prod_steps / mc_steps))  # type: ignore
+
+            if self.verbose:
+                self.logger.info("production phase complete")
+
+            if self.verbose:
+                self.logger.info("post-simulation result analysis")
```
Would advise against doing this here - we change logging level elsewhere in DAG execution, and we probably don't want the two competing.

IAlibay commented on openfe/protocols/openmm_septop/equil_septop_method.py:
```diff
@@ -0,0 +1,2125 @@
+          in either stateA or stateB.
+        """
+        # check that there is a protein component
+        if not any(
+            isinstance(comp, ProteinComponent) for comp in stateA.values()
+        ):
+            errmsg = "No ProteinComponent found in stateA"
+            raise ValueError(errmsg)
+
+        if not any(
+            isinstance(comp, ProteinComponent) for comp in stateB.values()
+        ):
+            errmsg = "No ProteinComponent found in stateB"
+            raise ValueError(errmsg)
+
+        # check that there is a solvent component
+        if not any(
+            isinstance(comp, SolventComponent) for comp in stateA.values()
+        ):
+            errmsg = "No SolventComponent found in stateA"
+            raise ValueError(errmsg)
+
+        if not any(
+            isinstance(comp, SolventComponent) for comp in stateB.values()
+        ):
+            errmsg = "No SolventComponent found in stateB"
+            raise ValueError(errmsg)
```
Sounds like something we maybe should move to gufe?

hannahbaumann replied:
Ok, I'll then leave it as is for now and make the changes after the Gufe level implementation.

IAlibay commented on openfe/protocols/openmm_septop/equil_septop_method.py:
```diff
@@ -0,0 +1,2125 @@
+        ------
+        ValueError
+          If the number of lambda windows differs for electrostatics and sterics.
+          If the number of replicas does not match the number of lambda windows.
```
This is a holdover of some unfinished work with SAMS. Technically you don't need the two to match, this is also true for unsampled endstates. We were meant to find time to try to loosen that restriction but never got around to it.

jthorton replied:
Ahh I see, I would still prefer that the task fail early we don't want users to submit the tasks to the HPC and wait in the queue only to instantly fail due to a wrong input setting, that's not a great experience. Can we enforce this in the settings for now or in the validate method of the protocol?

hannahbaumann commented on openfe/protocols/openmm_septop/equil_septop_method.py:
```diff
@@ -0,0 +1,2125 @@
+
+        # Crash out if there are less or more than one alchemical components
+        # in state A and B
+        if len(alchemical_components['stateA']) != 1:
+            errmsg = ("Exactly one alchemical components must be present in stateA. "
+                      f"Found {len(alchemical_components['stateA'])} "
+                      "alchemical components in stateA")
+            raise ValueError(errmsg)
+
+        if len(alchemical_components['stateB']) != 1:
+            errmsg = ("Exactly one alchemical components must be present in stateB. "
+                      f"Found {len(alchemical_components['stateB'])} "
+                      "alchemical components in stateB")
+            raise ValueError(errmsg)
```
Thanks! Changed this!

hannahbaumann commented on openfe/protocols/openmm_septop/equil_septop_method.py:
```diff
@@ -0,0 +1,2125 @@
+            raise ValueError(errmsg)
+
+        # Check if there are lambda windows with naked charges
+        for inx, lam in enumerate(lambda_elec_A):
+            if lam < 1 and lambda_vdw_A[inx] == 1:
+                errmsg = (
+                    "There are states along this lambda schedule "
+                    "where there are atoms with charges but no LJ "
+                    f"interactions: Ligand A: lambda {inx}: "
+                    f"elec {lam} vdW {lambda_vdw_A[inx]}")
+                raise ValueError(errmsg)
+            if lambda_elec_B[inx] < 1 and lambda_vdw_B[inx] == 1:
+                errmsg = (
+                    "There are states along this lambda schedule "
+                    "where there are atoms with charges but no LJ interactions"
+                    f": Ligand B: lambda {inx}: elec {lambda_elec_B[inx]}"
+                    f" vdW {lambda_vdw_B[inx]}")
+                raise ValueError(errmsg)
```
Thanks, changed this!

hannahbaumann commented on openfe/protocols/openmm_septop/base.py:
```diff
@@ -0,0 +1,1300 @@
+    topology: openmm.app.Topology,
+    positions: omm_unit.Quantity,
+    settings: dict[str, SettingsBaseModel],
+    endstate: str,
```
Changed this.

hannahbaumann commented:
Thanks @jthorton , I think I addressed all your comments, except the ones we will address in follow up PRs. Whenever you have some time this would be ready for re-review!

mikemhenry commented:
fix for the docs is here: https://github.com/hannahbaumann/openfe_septop/pull/4

atravitz commented on openfe/tests/protocols/openmm_septop/test_septop_slow.py:
```diff
@@ -0,0 +1,398 @@
+import pathlib
+import mdtraj as md
+
+from openfecli.tests.commands.test_atommapping import print_test
+
```
@hannahbaumann deleting this unused import will solve your mypy problem

hannahbaumann replied:
Awesome, thank you so much @atravitz , that fixed it!

IAlibay commented on openfe/protocols/openmm_septop/utils.py:
```diff
@@ -0,0 +1,165 @@
+
+    # Create parent directory if it doesn't exist
+    filename_basedir = filename.parent
+    if not filename_basedir.exists():
```
Probably just keep the path you're using. We can all agree to using bzip2 if that's the best.

hannahbaumann replied:
Do you mean I should just remove this part?

IAlibay commented:
Ran long tests today, got this:

FAILED openfe/tests/protocols/openmm_septop/test_septop_slow.py::test_openmm_run_engine[CUDA] - KeyError: 'last_checkpoint'

hannahbaumann commented:
@IAlibay I fixed the long test, it should be working now!

IAlibay commented on docs/guide/protocols/septop.rst:
```diff
@@ -0,0 +1,131 @@
+
+Orientational, or Boresch-style, restraints are automaticallly (unless manually specified) applied between three protein and three ligand atoms using one bond,
+two angle, and three dihedral restraints. Reference atoms are picked based on different criteria, such as the root mean squared
+fluctuation of the atoms in a short MD simulation, the secondary structure of the protein, and the distance between atoms, based on heuristics from Baumann et al. [2]_.
```
TODO (issued raised): update the docs here (and in ABFEs) to reflect the new picking, i.e. it's getting closer to a mix between Hannah's appproach and MDRestraintsGenerator.

IAlibay commented on docs/guide/protocols/septop.rst:
```diff
@@ -0,0 +1,131 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In the :class:`SepTopProtocol <.SepTopProtocol>` the coulombic interactions of the molecules are fully turned off (annihilated) in the respective non-interacting end states.
+The Lennard-Jones interactions are instead decoupled, meaning the intermolecular interactions turned off, keeping the intramolecular Lennard-Jones interactions.
```
```suggestion
The Lennard-Jones interactions are instead decoupled, meaning the intermolecular interactions are turned off, keeping the intramolecular Lennard-Jones interactions.
```

IAlibay commented on docs/guide/protocols/septop.rst:
```diff
@@ -0,0 +1,131 @@
+Molecular interactions are modified along an alchemical path using a discrete set of lambda windows.
+For the transformation of ligand A to ligand B in the binding site, the following steps are carried out, starting with ligand A being fully interacting in the binding site while ligand B is decoupled.
+
+1. Insert the non-interacting dummy ligand B into the binding site and restrain it using orientational restraints. The contribution of the restraints is calculated analytically.
```
Separate issue raised: the description of the lambda schedule is non symmetric, i.e. number 5 describes transferring the ligand A to solvent, but none of these refer to transferring ligand B from the solvent (I think it's what is meant by "insert the non-interacting dummy", but the language probably needs to be improved).

IAlibay commented on docs/guide/protocols/septop.rst:
```diff
@@ -0,0 +1,131 @@
+
+The :class:`.ProtocolDAG` of the :class:`SepTopProtocol <.SepTopProtocol>` contains :class:`.ProtocolUnit`\ s from both the complex and solvent transformations.
+This means that both legs of the thermodynamic cycle are constructed and run sequentially in the same :class:`.ProtocolDAG`. This is different from the :class:`.RelativeHybridTopologyProtocol` where the :class:`.ProtocolDAG` only runs a single leg of a thermodynamic cycle.
+If multiple ``protocol_repeats`` are run (default: ``protocol_repeats=3``), the :class:`.ProtocolDAG` contains multiple :class:`.ProtocolUnit`\ s of both complex and solvent transformations.
```
Unclear from this text if that means that the same restraint will be used for all repeats.

IAlibay commented on docs/guide/protocols/septop.rst:
```diff
@@ -0,0 +1,131 @@
+Here are some details of how the simulation is carried out which are not detailed in the :class:`SepTopProtocol <.SepTopProtocol>`:
+
+* The protocol applies a `LangevinMiddleIntegrator <https://openmmtools.readthedocs.io/en/latest/api/generated/openmmtools.mcmc.LangevinDynamicsMove.html>`_ which uses Langevin dynamics, with the LFMiddle discretization [4]_.
+* A MonteCarloBarostat is used in the NPT ensemble to maintain constant pressure.
```
```suggestion
* A `Monte Carlo barostat <https://docs.openmm.org/latest/api-python/generated/openmm.openmm.MonteCarloBarostat.html>`_ is used in the NPT ensemble to maintain constant pressure.
```

IAlibay reviewed (commented):
review for tests

IAlibay commented on openfe/tests/protocols/openmm_septop/test_septop_protocol.py:
```diff
@@ -0,0 +1,1411 @@
+    # Fixing the number of solvent molecules in the solvent settings
+    # to test against reference xml
+    default_settings.solvent_solvation_settings.solvent_padding = None
+    default_settings.solvent_solvation_settings.number_of_solvent_molecules = 364
```
```suggestion
    default_settings.solvent_solvation_settings.number_of_solvent_molecules = 364
    default_settings.forcefield_settings.small_molecule_forcefield = "openff-2.1.1"
```

IAlibay commented on openfe/tests/protocols/openmm_septop/test_septop_protocol.py:
```diff
@@ -0,0 +1,1411 @@
+
+        for a, b in zip(constraints, constraints_ref):
+            # Particle 1
+            assert pytest.approx(a[0].value_in_unit(openmm.unit.nm), rel=1e-4) == b[0].value_in_unit(openmm.unit.nm)
+            # Particle 2
+            assert pytest.approx(a[1].value_in_unit(openmm.unit.nm), rel=1e-4) == b[1].value_in_unit(openmm.unit.nm)
+            # Constraint Quantity
+            assert pytest.approx(a[2].value_in_unit(openmm.unit.nm), rel=1e-4) == b[2].value_in_unit(openmm.unit.nm)
```
```suggestion
            assert a[0] == b[0]
            # Particle 2
            assert a[1] == b[1]
            # Constraint Quantity
            assert a[2] == b[2]
```

IAlibay commented on openfe/protocols/openmm_septop/equil_septop_method.py:
```diff
@@ -0,0 +1,2125 @@
+        )
+        rdmol_A = mol_A.to_rdkit()
+        rdmol_B = mol_B.to_rdkit()
+        Chem.SanitizeMol(rdmol_A)
+        Chem.SanitizeMol(rdmol_B)
```
Opened issue #1526 - we should document at the very least.

IAlibay commented on openfe/protocols/openmm_septop/equil_septop_method.py:
```diff
@@ -0,0 +1,2338 @@
+
+        # 2. Get settings
+        settings = self._handle_settings()
+        print(settings)
```
```suggestion
```
